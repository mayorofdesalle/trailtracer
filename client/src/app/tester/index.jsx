import '@testing-library/jest-dom';

import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';

import './mocks/i18n.mock';
import './mocks/matchMedia.mock';
import './mocks/resizeObserver.mock';

import store from '@app/store';
import GlobalStyle from '@components/style/GlobalStyle';
import Theme from '@components/style/Theme';

const Providers = ({ children }) => {
    return (
        <Provider store={store}>
            <Theme>
                <GlobalStyle />
                <BrowserRouter>
                    {children}
                </BrowserRouter>
            </Theme>
        </Provider>
    );
};

const customRender = (ui, options) =>
    render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';
export * from 'vitest';

// override render method
export { customRender as render };
