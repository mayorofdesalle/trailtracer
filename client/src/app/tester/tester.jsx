import { render } from '@testing-library/react'
import { Provider } from 'react-redux';

import './mocks/matchMedia.mock';
import store from '@app/store';
import Theme from '@components/style/Theme'

const Providers = ({ children }) => {
    return (
        <Provider store={store}>
            <Theme>
                {children}
            </Theme>
        </Provider>
    );
};

const customRender = (ui, options) =>
    render(ui, { wrapper: Providers, ...options });

// re-export everything
export * from '@testing-library/react';
export * from 'vitest';

// override render method
export { customRender as render };