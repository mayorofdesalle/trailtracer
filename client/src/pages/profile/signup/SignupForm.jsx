import { useMemo, useState, useCallback, useEffect } from 'react';
import { useForm } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import styled, { useTheme } from 'styled-components';

import { logger } from '@app';
import { Button, Container, Icon, Text } from '@components/ui';
import { PasswordInput, TextInput, changeFocus } from '@components/ui/input';
import { FormContext } from '@context';

const SignupFormInner = styled.form`
    position: relative;
    display: flex;
    flex-direction: column;
    justify-content: space-around;
    align-items: center;
    
    & > * {
        margin-top: clamp(1rem, min(3dvw, 3dvh), 3rem);
    }

    & > ${Container}:first-of-type {
        margin-top: 0;
    }

    & > ${Button} {
        padding: 0.25rem;
        width: 100%;
    }
`;

const InputContainer = styled(Container)`
    position: relative;
    flex-direction: column;
    align-items: start;
    justify-content: space-around;

    & > ${Text} {
        margin-left: clamp(0.5rem, min(1dvw, 1dvh), 1rem);
    }

    & input {

        &::placeholder {
            color: transparent;
        }
    }
`;

/**
 * SignupForm
 * @description
 * Form to sign in with email and password.
 * **/
const SignupForm = (props) => {
    const theme = useTheme();
    const { t } = useTranslation();

    const { register, handleSubmit, setFocus, setValue, watch, formState: { errors } } = useForm({
        defaultValues: useMemo(() => {
            return {
                email: '',
                'new-password': ''
            };
        }, []),
        mode: 'onBlur',
        reValidateMode: 'onBlur',
        shouldFocusError: true
    });

    const [isSubmitting, setIsSubmitting] = useState(false);

    const onSubmit = useCallback((data) => {
        setIsSubmitting(true);
        logger.debug(data);
    });

    useEffect(() => {
        setFocus('email');
    }, []);

    return (
        <FormContext.Provider value={{ register, setFocus, setValue, watch }}>
            <SignupFormInner role='form' onSubmit={handleSubmit(onSubmit)} onKeyDown={(e) => changeFocus(e, ['email', 'new-password'], setFocus)} {...props}>
                <InputContainer>
                    <Text $color={theme.colors.primary} $ratio={0.75}>{t('forms.email')}</Text>
                    <TextInput type='email' error={errors.email} required validate>
                        <Icon name='at-fill' color={errors.email ? theme.colors.secondary : theme.colors.primary} />
                    </TextInput>
                </InputContainer>
                <InputContainer>
                    <Text $color={theme.colors.primary} $ratio={0.75}>{t('forms.new-password')}</Text>
                    <PasswordInput error={errors['new-password']} name='new-password' required validate>
                        <Icon name='key-fill' color={errors['new-password'] ? theme.colors.secondary : theme.colors.primary} />
                    </PasswordInput>
                </InputContainer>
                <Button $type='submit' $bgColor={theme.colors.primary} disabled={isSubmitting}>
                    <Text $color={theme.colors.background} $heading>{t('forms.continue')}</Text>
                </Button>
            </SignupFormInner>
        </FormContext.Provider>
    );
};

export default SignupForm;