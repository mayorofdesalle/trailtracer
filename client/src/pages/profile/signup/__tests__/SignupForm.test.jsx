import { describe, expect, it, render, screen } from '@app/tester';

import SignupForm from '../SignupForm';

describe('SignupForm', () => {

    it('should render correctly', () => {
        render(<SignupForm />);

        const form = screen.getByRole('form');
        expect(form).toBeInTheDocument();

        const email = screen.getByRole('textbox');
        expect(email).toBeInTheDocument();

        const password = screen.getByLabelText(/password/i);
        expect(password).toBeInTheDocument();

        const button = screen.getByRole('button');
        expect(button).toBeInTheDocument();
    });

});