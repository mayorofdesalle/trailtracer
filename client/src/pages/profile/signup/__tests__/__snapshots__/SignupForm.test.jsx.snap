// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SignupForm > should match snapshot 1`] = `
<DocumentFragment>
  <form
    class="sc-eTNRI kZqNyz"
    role="form"
  >
    <div
      class="sc-guDLey sc-brSamD kHwwsn hxA-Dbk"
    >
      <div
        class="sc-gLLuof jmevau"
      >
        forms.email
      </div>
      <div
        class="sc-guDLey sc-irLvIq kHwwsn bHzjpc"
      >
        <label
          for="email"
          hidden=""
        >
          forms.email
        </label>
        <input
          autocomplete="email"
          class="sc-kFCroH kdoZoe"
          data-tooltip-id="email"
          id="email"
          name="email"
          placeholder="forms.email"
          type="email"
        />
        <svg
          class="sc-eDLKkx kBnHHP"
          fill="#435049"
          role="img"
        >
          <title>
            at-fill
          </title>
          <use
            href="http://localhost:3000/src/assets/icons.svg#at-fill"
          />
        </svg>
      </div>
    </div>
    <div
      class="sc-guDLey sc-brSamD kHwwsn hxA-Dbk"
    >
      <div
        class="sc-gLLuof jmevau"
      >
        forms.new-password
      </div>
      <div
        class="sc-guDLey sc-irLvIq kHwwsn bHzjpc"
      >
        <label
          for="password"
          hidden=""
        >
          forms.new-password
        </label>
        <input
          autocomplete="new-password"
          class="sc-kFCroH gCiziP"
          data-tooltip-id="password"
          id="password"
          name="new-password"
          placeholder="forms.new-password"
          type="password"
        />
        <svg
          class="sc-eDLKkx kBnHHP"
          fill="#435049"
          role="img"
        >
          <title>
            key-fill
          </title>
          <use
            href="http://localhost:3000/src/assets/icons.svg#key-fill"
          />
        </svg>
      </div>
    </div>
    <button
      class="sc-beySPh zbkUW"
      type="submit"
    >
      <div
        class="sc-gLLuof ldUHqL"
      >
        forms.continue
      </div>
    </button>
  </form>
</DocumentFragment>
`;
